File: lesson11-more-bytecode/src/main/java/pt/isel/leic/lae/li42d/lesson11/ex03/Prog.java
package pt.isel.leic.lae.li42d.lesson11.ex03;
interface I {}
class B implements I {}
final class D extends B {}

public class Prog {

	static Object  upcast1(Integer value)  { return value; }           // no checkcast
	static I       upcast2(B value)        { return value; }           // no checkcast
	static B       upcast3(D value)        { return value; }           // no checkcast

	static Integer downcast1(Object value) { return (Integer)value; }     // checkcast
	static B       downcast2(I value)      { return       (B)value; }     // checkcast
	static D       downcast3(B value)      { return       (D)value; }     // checkcast

	static <T> T   identity(T value)       { return value; }           // no checkcast

	// due to type erasure (JVM does not support generics), all usages
	// of generic functions/methods require casts of output values
	static Integer passThrough(Integer value) { return identity(value); } // checkcast
	static String  passThrough(String  value) { return identity(value); } // checkcast

	public static void main(String[] args) {
		final var u1 = upcast1(1234);
		final var u2 = upcast2(new B());
		final var u3 = upcast3(new D());

		final var d1 = downcast1(1234);
		final var d2 = downcast2(new B());
		final var d3 = downcast3(new D());

		final var i1 = passThrough(1234);
		final var i2 = passThrough("ISEL");

		System.out.println(u1);
		System.out.println(u2);
		System.out.println(u3);

		System.out.println(d1);
		System.out.println(d2);
		System.out.println(d3);

		System.out.println(i1);
		System.out.println(i2);
	}
}

File: lesson11-more-bytecode/src/main/kotlin/ex03-casts/prog.kt
package pt.isel.leic.lae.li42d.lesson11.ex03_casts


private interface I {}
private open class B() : I {}
private class D() : B() {}

private fun upcast1(value: Int?) : Any? = value    // no checkcast
private fun upcast2(value: B?) : I? = value        // checkcast in upcast?
private fun upcast3(value: D?) : B? = value        // checkcast in upcast?

// all downcasts must be explicit and generate checkcast instructions
private fun downcast1(value: Any?) : Int? = value as Int?   
private fun downcast2(value: I?) : B? = value as B?
private fun downcast3(value: B?) : D? = value as D?

private fun <T> identity(value: T?) : T? = value   // no checkcast

// due to type erasure (JVM does not support generics), all usages
// of generic functions/methods require casts of output values
private fun passThrough(value: Int?)    : Int?    = identity(value)  // checkcast
private fun passThrough(value: String?) : String? = identity(value)  // checkcast

fun main() {
	val u1 = upcast1(1234)
	val u2 = upcast2(B())
	val u3 = upcast3(D())

	val d1 = downcast1(1234)
	val d2 = downcast2(B())
	val d3 = downcast3(D())

	val i1 = passThrough(1234)
	val i2 = passThrough("ISEL")

	println(u1)
	println(u2)
	println(u3)

	println(d1)
	println(d2)
	println(d3)

	println(i1)
	println(i2)
}
